tonic/datasets/slicers.py

```diff
@@ -0,0 +1,262 @@
+from dataclasses import dataclass
+from typing_extensions import Protocol
+from typing import Any, List, Tuple
+
+import numpy as np
+
+
+class Slicer(Protocol):
+    def get_slice_metadata(self, data: Any) -> List[Any]:
+        """
+        This method returns the meta data that helps with slicing of the given data.
+        Eg. it could return the indices at which the data would be sliced.
+        The return value of this method should be usable by the method slice_with_meta.
+
+        Args:
+            data:
+
+        Returns:
+            metadata
+        """
+        ...
+
+    def slice_with_metadata(self, data: Any, metadata: Any) -> List[Any]:
+        """
+        Slice the data using the metadata.
+        Args:
+            data:
+            metadata:
+
+        Returns:
+            sliced_data
+        """
+        ...
+
+    def slice(self, data: Any) -> List[Any]:
+        """
+        Slice the given data and return the sliced data
+        Args:
+            data:
+
+        Returns:
+            sliced_data
+        """
+        ...
+
+
+@dataclass(frozen=True)
+class SliceByTime:
+    """
+    Return xytp split according to fixed timewindow and overlap size
+    <        <overlap>        >
+    |   window1      |
+             |   window2      |
+
+    Args:
+        time_window: int
+            Length of time for each xytp (ms)
+        overlap: int
+            Length of time of overlapping (ms)
+        include_incomplete: bool
+            include incomplete slices ie potentially the last xytp
+    """
+    time_window: float
+    overlap: float = 0.0
+    include_incomplete: bool = False
+
+    def slice(self, data: np.ndarray) -> List[np.ndarray]:
+        metadata = self.get_slice_metadata(data)
+        return self.slice_with_metadata(data, metadata)
+
+    def get_slice_metadata(self, data: np.ndarray) -> List[Tuple[int, int]]:
+        t = data["t"]
+        stride = self.time_window - self.overlap
+        assert stride > 0
+
+        if self.include_incomplete:
+            n_slices = int(np.ceil(((t[-1] - t[0]) - self.time_window) / stride) + 1)
+        else:
+            n_slices = int(np.floor(((t[-1] - t[0]) - self.time_window) / stride) + 1)
+        n_slices = max(n_slices, 1)  # for strides larger than recording time
+
+        tw_start = np.arange(n_slices) * stride + t[0]
+        tw_end = tw_start + self.time_window
+        indices_start = np.searchsorted(t, tw_start)[:n_slices]
+        indices_end = np.searchsorted(t, tw_end)[:n_slices]
+        return list(zip(indices_start, indices_end))
+
+    @classmethod
```

Comment:
If I'm not mistaken, this should be a staticmethod instead

Reply:
Static methods are not overloadable i suspect and in any case are not used in python most often.

Reply:
I assume it doesn't make a difference what these methods are tagged as since it is unlikely they will be inherrited/overriden to define a custom slicer.